File: Backend/TopSys.TopConWeb.Domain/Services/LeadService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Topsys.TopConWeb.SharedKernel.Common;
using TopSys.TopConWeb.Domain.Entities.Lead;
using TopSys.TopConWeb.Domain.Interfaces.Repositories;
using TopSys.TopConWeb.Domain.Interfaces.Services;
using TopSys.TopConWeb.SharedKernel.Helpers;

namespace TopSys.TopConWeb.Domain.Services
{
    public class LeadService : ServiceBase<Lead>, ILeadService
    {
        private readonly ILeadRepository _leadRepository;
        private readonly ILeadContatoRepository _leadContatoRepository;

        public LeadService(ILeadRepository leadRepository, ILeadContatoRepository leadContatoRepository) : base(leadRepository)
        {
            _leadRepository = leadRepository;
            _leadContatoRepository = leadContatoRepository;
        }

        public void Adicionar(string usuario, Lead lead)
        {
            const int USINA_DEFAULT = 999;

            lead.IdCadastro = StringHelper.GetIDD(usuario);
            lead.UsinaCodigo = USINA_DEFAULT;
            lead.Ano = int.Parse($"{lead.Data.Year}".Substring(2));

            lead.ContatoPrincipal.Usina = lead.UsinaCodigo;
            lead.ContatoPrincipal.AnoLead = lead.Ano;
            lead.ContatoPrincipal.Sequencia = 1;

            if (lead.ContatoSecundario != null)
            {
                lead.ContatoSecundario.Usina = lead.UsinaCodigo;
                lead.ContatoSecundario.AnoLead = lead.Ano;
                lead.ContatoSecundario.Sequencia = 2;
            }

            if (lead.Classificacao == 0)
                lead.Classificacao = Enums.EClassificacaoTemperatura.Frio;

            _leadRepository.Adicionar(lead);

            lead.ContatoPrincipal.NumeroLead = lead.Numero;
            _leadContatoRepository.Adicionar(lead.ContatoPrincipal);

            if (lead.ContatoSecundario != null)
            {
                lead.ContatoSecundario.NumeroLead = lead.Numero;
                _leadContatoRepository.Adicionar(lead.ContatoSecundario); 
            }
        }

        public void AdicionarAnexo(string usuario, Guid id, int usina, int anoLead, int numeroLead, string anexo, string nome)
        {
            _leadRepository.AdicionarAnexo(usuario, id, usina, anoLead, numeroLead, anexo, nome);
        }

        public ICollection<LeadAnexo> ListarAnexos(int usina, int anoLead, int numeroLead)
        {
            return _leadRepository.ListarAnexos(usina, anoLead, numeroLead);
        }

        public byte[] ObterAnexo(Guid id)
        {
            return _leadRepository.ObterAnexo(id);
        }

        public LeadAnexo ObterLeadAnexoPorId(Guid id)
        {
            return _leadRepository.ObterLeadAnexoPorId(id);
        }

        public void AtualizarDescricaoAnexo(LeadAnexo anexo)
        {
            _leadRepository.AtualizarDescricaoAnexo(anexo);
        }

        public void RemoverAnexo(Guid id)
        {
            _leadRepository.RemoverAnexo(id);
        }

        public PagedList<Lead> ListarEmOrdemDecrescente(int pagina, int porPagina, Expression<Func<Lead, bool>> filter)
        {
            var pagedList = _leadRepository.ListarEmOrdemDecrescente(pagina, porPagina, filter);

            return pagedList;
        }

        public Lead ObterPorUsinaAnoNumero(int idUsina, int ano, int numero, bool tracking = false)
        {
            var lead = _leadRepository.ObterPorUsinaAnoNumero(idUsina, ano, numero, tracking);

            return lead;
        }

        public PagedList<LeadInteracao> ListarInteracoes(int pagina, int porPagina, Expression<Func<LeadInteracao, bool>> filter)
        {
            return _leadRepository.ListarInteracoes(pagina, porPagina, filter);
        }

        public void AdicionarInteracao(string usuario, LeadInteracao leadInteracao)
        {
            _leadRepository.AdicionarInteracao(usuario, leadInteracao);
        }
    }
}

File: Frontend/TopConWeb-AngularV8/src/app/services/lead.service.ts
import { Injectable, Injector } from '@angular/core';

import { BaseService } from './base.service';
import { Lead } from 'app/classes/lead/lead';
import { Usina } from 'app/classes/usina/usina';
import { LeadLog } from 'app/classes/lead/lead-log';
import { LeadAnexo } from 'app/classes/lead/lead-anexo';
import { LeadInteracao } from 'app/classes/lead/lead-interacao';
import { Oportunidade } from 'app/classes/oportunidade/oportunidade';

@Injectable()
export class LeadService extends BaseService {
    constructor(injector: Injector) {
        super(injector);
    }

    Adicionar(lead: Lead) {
        return this.makePostPrommisse<Lead>('v1/lead', JSON.stringify(lead));
    }

    Atualizar(lead: Lead) {
        return this.makePatchPrommisse<any>('v1/lead', JSON.stringify(lead));
    }

    ListarEmOrdemDecrescente(pagina: number, porPagina: number, filtro?: string, hideLoading?: boolean) {
        return this.makePagedGetPrommisse<Lead>('v1/leads?'
            +'pagina='+pagina
            +'&porPagina='+porPagina
            +(filtro ? '&'+filtro : '')
            , hideLoading);
    }

    ObterPorUsinaAnoNumero(usina: Usina, ano: number, numero: number, hideLoading?: boolean){
        return this.makeGetPrommisse<Lead>('v1/lead'
            + '/usina/' + usina.codigo
            + '/ano/' + ano
            + '/numero/' + numero,
            hideLoading);
    }

    obterOportunidadeDeLead(usina: number, ano: number, numero: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<Oportunidade>(`v1/lead/usina/${usina}/ano/${ano}/numero/${numero}/gerar-oportunidade`, hideLoading);
    }

    ListarLeadLogsPorId(usina: number, ano: number, numero: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<LeadLog[]>('v1/lead'
            + '/usina/' + usina
            + '/ano/' + ano
            + '/numero/' + numero
            + '/logs/',
            hideLoading);
    }

    adicionarAnexo(usina: number, anoLead: number, numeroLead: number, anexo: string, nomeAnexo: string, hideLoading?: boolean) {
        var body = {
            usina: usina,
            ano: anoLead,
            numero: numeroLead,
            arquivo: anexo,
            nome: nomeAnexo
        };
        return this.makePostPrommisse<string>('v1/lead/anexo', JSON.stringify(body), hideLoading);
    }

    listarAnexos(usina: number, anoLead: number, numeroLead: number, hideLoading?: boolean) {
        return this.makeGetPrommisse<LeadAnexo[]>('v1/lead/anexo'
        +'/usina/' + usina
        +'/ano/'+ anoLead
        +'/numero/' + numeroLead
        , hideLoading);
    }

    ObterAnexo(anexoId: string, hideLoading?: boolean){
        return this.makeGetPrommisse<string>('v1/lead/anexo/' + anexoId, hideLoading);
    }

    atualizarDescricaoAnexo(anexo: LeadAnexo) {
        return this.makePatchPrommisse<any>('v1/lead/anexo', JSON.stringify(anexo));
    }

    removerAnexo(id: string, hideLoading?: boolean) {
        return this.makeDeletePrommisse<LeadAnexo>('v1/lead/anexo/' + id, hideLoading);
    }

    adicionarInteracao(leadInteracao: LeadInteracao, hideLoading?: boolean) {
        return this.makePostPrommisse<string>('v1/lead/interacao', JSON.stringify(leadInteracao), hideLoading);
    }

    listarInteracoes(pagina: number, porPagina: number, filtro?: string, hideLoading?: boolean) {
        return this.makePagedGetPrommisse<LeadInteracao>('v1/lead/interacoes?'
            +'pagina='+pagina
            +'&porPagina='+porPagina
            +(filtro ? '&'+filtro : '')
            , hideLoading);
    }
}